src/metabase/lib/join.cljc

```diff
@@ -1,4 +1,19 @@
 (ns metabase.lib.join
+  "Functions related to manipulating EXPLICIT joins in MBQL.
+
+  Several of the functions in this namespace operate on a 'joinable'. This means anything that we can join:
+
+  * A join map, i.e. a `:lib/type` `:mbql/join` map
+
+  * A Table (a `:metadata/table` map)
+
+  * A Saved Question/Card (a `:metadata/card` map)
+
+  * Another query (an `:mbql/query` map)
+
+  Realistically this is probably more flexibility than we actually needed, so maybe we can pare this stuff down a bit
+  and only have things like [[with-join-alias]] operate on join maps going forward. But that's a project for another
+  day."
```

Comment:
I'm going back and forth between just cleaning this stuff all up now and simplifying everything or opening a GH issue and doing it later. But I think having stuff like `with-join-alias` work for TableMetadatas and CardMetadatas and Joins and functions and `:field` clauses and ColumnMetadatas is just preposterous, I don't think it's really that unreasonable to just ask people to call `join-clause` first and then call the other functions.

Reply:
Opened https://github.com/metabase/metabase/issues/31678 to track this